### DISCUSSION TITLE
Release Notes: v2.35.0

### DISCUSSION THREAD

**Discussion by PavelBlend** [PavelBlend/blender-xray#721] [General]
## Улучшения:
- Добавлена поддержка импорта/экспорта `*.group` формата (#673):

Оператор импорта:
![01](https://github.com/PavelBlend/blender-xray/assets/7983249/5e16c4e0-e6b1-4fc7-8d90-6a04a111532a)

Оператор экспорта:
![02](https://github.com/PavelBlend/blender-xray/assets/7983249/911760d2-6df3-49f7-9e5e-2f00bfa6636b)

N-панель:
![05](https://github.com/PavelBlend/blender-xray/assets/7983249/3553002c-4ea3-40fc-bc59-7fbb4464b208)

Значения параметров импорта по-умолчанию:
![03](https://github.com/PavelBlend/blender-xray/assets/7983249/4e195134-93f1-47fe-a378-046131afa470)

Галочки включения/отключения импорта/экспорта формата `group`:
![04](https://github.com/PavelBlend/blender-xray/assets/7983249/4032412b-507a-4901-9a23-96f8bfe50abe)

При импорте `group`, аддон создаёт коллекцию, имя которой совпадает с именем файла:
![06](https://github.com/PavelBlend/blender-xray/assets/7983249/9f017e21-9050-4e71-b2d3-4f3646ba9694)

Чтобы сделать экспорт, нужно выделить объекты из коллекции с помощью ПКМ:
![07](https://github.com/PavelBlend/blender-xray/assets/7983249/6253c697-d0e9-4b62-87ea-509f21f26009)

При экспорте в `group`, в файл попадут выделенные root-объекты. Относительный путь вычисляется так же, как и для `scene selection` экспорта, то есть по `Export Path` root-объекта.

- Добавлена поддержка экспорта `*.part` файлов в двоичном формате ТЧ (#494). Поддерживается экспорт только поверх существующего файла, так как все `part` файлы локации (`scene_object.part`, `group.part`, `ai_map.part` и т. д.) должны иметь одинаковые данные, которые указывают, что эти `part` файлы от одного и того же уровня. Если экспортировать в новый файл, то будет ошибка:
![09](https://github.com/PavelBlend/blender-xray/assets/7983249/e0fe4638-02e7-43b2-8557-6e7dc4251610)

Если готового `part` файла нет, то можно создать пустую локацию и сохранить её.

- При импорте `part` файлов, как и для `group`, то же создаётся коллекция:
![08](https://github.com/PavelBlend/blender-xray/assets/7983249/a971e01b-150b-43d7-b759-82d297b06220)

- При импорте `game level`, источникам света указывается тип и цвет:
![12](https://github.com/PavelBlend/blender-xray/assets/7983249/9dae7083-abf4-47e7-b409-457af95d38f1)

Источники света теперь имеют такое влияние:
![10](https://github.com/PavelBlend/blender-xray/assets/7983249/7e788e12-b9df-4601-90c9-fea9004da218)

Вместо того, что было в предыдущей версии аддона:
![11](https://github.com/PavelBlend/blender-xray/assets/7983249/04185734-2452-4173-a76b-00b83915654c)

## Исправления:
- Исправлен импорт цвета от источников света из `game level`. Теперь к цвету применяется гамма коррекция:
![light_color](https://github.com/PavelBlend/blender-xray/assets/7983249/8a491364-cd92-4550-8684-442504898b97)

**Comment by kusyukin** [comment]
Павел, здравствуйте! Столкнулся с такой проблемой при импорте игровых моделей.
Можете пожалуйста подсказать, в чем дело?

```
Python: Traceback (most recent call last):
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\log.py", line 284, in wrapper
    return method(self, context)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\stats.py", line 277, in wrapper
    result = method(self, context)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\ie.py", line 70, in wrapper
    result = method(self, context, *args)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\ops.py", line 88, in execute
    utils.ie.import_files(
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\ie.py", line 200, in import_files
    report_errors(context)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\ie.py", line 215, in report_errors
    raise first_error
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\ie.py", line 191, in import_files
    result = imp_fun(file_path, context)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\log.py", line 270, in wrap
    return func(*args, **kwargs)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\stats.py", line 238, in wrapper
    result = method(*args, **kwargs)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\main.py", line 117, in import_file
    import_ogf_visual(context, data, visual)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\main.py", line 61, in import_ogf_visual
    import_fun(context, chunks, ogf_chunks, visual)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\load.py", line 79, in read_mt_skeleton_anim
    read_mt_skeleton_rigid(context, chunks, ogf_chunks, visual)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\load.py", line 74, in read_mt_skeleton_rigid
    child.import_children(context, chunks, ogf_chunks, visual)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\child.py", line 58, in import_children
    main.import_ogf_visual(context, child_data, visual)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\main.py", line 61, in import_ogf_visual
    import_fun(context, chunks, ogf_chunks, visual)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\load.py", line 84, in read_mt_skeleton_geom_def_st
    shader.read_texture(context, chunks, ogf_chunks, visual)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\formats\ogf\imp\shader.py", line 25, in read_texture
    bpy_material, bpy_image = utils.material.get_material(
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\material.py", line 336, in get_material
    bpy_material, bpy_image = _create_material_and_image(
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\material.py", line 290, in _create_material_and_image
    princ_node = create_mat_nodes(bpy_material)
  File "C:\Users\Kiril\AppData\Roaming\Blender Foundation\Blender\4.0\scripts\addons\io_scene_xray\utils\material.py", line 183, in create_mat_nodes
    princ_node.inputs['Specular'].default_value = 0.0
KeyError: 'bpy_prop_collection[key]: key "Specular" not found'
bpy.ops.xray_import.ogf(filepath="D:\\S.T.A.L.K.E.R Anomaly\\tools\\_unpacked\\meshes\\actors\\stalker_monolith\\stalker_monolith1_gp5.ogf", directory="D:\\S.T.A.L.K.E.R Anomaly\\tools\\_unpacked\\meshes\\actors\\stalker_monolith\\", files=[{"name":"stalker_monolith1_gp5.ogf", "name":"stalker_monolith1_gp5.ogf"}], import_motions=True, processed=False)

```

**Comment by PavelBlend** [comment]
@arenacreatoru привет. Аддон не будет работать в блендере 4.0

**Reply by kusyukin** [reply]
Понял, благодарю!

**Reply by PavelBlend** [reply]
3.4.1 Вот подробнее: #715

**Reply by kusyukin** [reply]
Павел, еще один вопрос есть. Я установил версию 2.93 и столкнулся со следующей проблемой. При импорте локаций, не все текстуры прогружаются. Как это можно исправить? 
![image](https://github.com/PavelBlend/blender-xray/assets/136234486/c1cae3dc-135a-4d5d-9cf8-21597ca6d97a)

**Comment by PavelBlend** [comment]
@arenacreatoru белые полигоны - это коллизия (level.cform), которая не имеет текстур. Они и должны быть белыми.

**Reply by kusyukin** [reply]
Подскажите пожалуйста, есть ли возможность с вами как-нибудь связаться не через гитхаб?

**Reply by kusyukin** [reply]
![image](https://github.com/PavelBlend/blender-xray/assets/136234486/979609b1-c1ad-4fe9-9528-8107cc6c43c6)
Проблема в том, что текстуры загружает как-то некорректно. Хотел у вас поинтересоваться, как это можно исправить?

**Reply by PavelBlend** [reply]
@arenacreatoru ну если некорректно загружает текстуры, то создай issue.

По скриншоту вроде бы всё правильно выглядит. Кусок террейна имеет материал с нодой `terrain\terrain_escape`. Может в моде текстура террейна такая серая изначально.

И на террейне не должно быть текстуры травы, так как трава накладывается с помощью шейдеров. Текстура травы много раз повторяется, создавая детализацию и смешивается с основной текстурой террейна. Иначе пришлось бы текстуру террейна делать не 1024x1024, а 100000x100000.

**Reply by kusyukin** [reply]
Спасибо большое! Постараюсь разобраться

**Reply by kusyukin** [reply]
Павел, здравствуйте. Подскажите пожалуйста, как текстуру террейна заменить на нормальные цветные текстуры? Это все вручную нужно делать?

**Reply by PavelBlend** [reply]
@arenacreatoru привет. Из какого мода локации? Может в моде всё так настроено? Посмотри как текстура выглядит. Если она цветная, то покажи, как выглядят ноды материала. Может ноды не правильно соединились при импорте. Хотя вряд ли.

И посмотри какой путь к файлу текстуры. Может пути не правильно вычислились при импорте:
![01](https://github.com/PavelBlend/blender-xray/assets/7983249/7dbe2567-c252-4540-939a-33049238e96a)

И проверь пути в настройках аддона. Может они ссылаются не на ту textures папку.

Другие варианты пока не приходят в голову.

**Reply by kusyukin** [reply]
![image](https://github.com/PavelBlend/blender-xray/assets/136234486/4acc4f5f-d8aa-41e4-8108-e3c9e0c8ec01)
Оказывается текстура из мода сама по себе ЧБ. Нужно будет поискать цветную. Я изначально думал, что ЧБ террейн - изначально запланированная вами фича. Пытался как-то в функциях тулса это пофиксить, но ничего не выходило. Большое спасибо за помощь! 
Локация из модификации Anomaly.

**Reply by kusyukin** [reply]
Случаем по вашему инструменту нет полноценных гайдов? Как допустим локацию травой наполнить с помощью шейдеров (вы чуть выше писали об этом)

**Reply by PavelBlend** [reply]
@arenacreatoru по траве урока нет. Но на ютубе есть канал с другими уроками по сталкеру. Канал Pavel_Blend. Там есть пару плейлистов с уроками по сталкеру.

Траву можно наложить таким способом:
![01](https://github.com/PavelBlend/blender-xray/assets/7983249/88e5ee7d-916c-4aab-ae5b-4f1dfe097f1e)

Только зачем траву добавлять? Экспортёр всё равно её не будет учитывать. Разве что для рендеринга видео добавлять её.

**Reply by kusyukin** [reply]
Спасибо большое! Мне для рендера картинок нужна локация, наполненная травой.

**Comment by Max1m1661** [comment]
Здравствуйте, столкнулся с проблемой экспорта .ogf

Модель деформируется при экспорте, даже если просто сделать Импорт -> Экспорт не меняя ничего, всё равно на выходе модель деформируется 

Модель до Экспорта
![11-1](https://github.com/PavelBlend/blender-xray/assets/138521479/9835e2c0-7597-4e4b-953b-bc91dba3e615)

И после
![11-2](https://github.com/PavelBlend/blender-xray/assets/138521479/121dd641-0e76-460e-b5b0-a20e9e7a8a00)

Анимация чисто для наглядности, обычные модели из ЗП тоже деформируются, но на этих деформация нагляднее...

**Comment by PavelBlend** [comment]
@Max1m1661 скорее всего это связано с весами вершин для костей. В формате ЧН/ЗП поддерживается до 4 весов на одну вершину. Но в ТЧ максимальное количество весов на вершину равно 2. Аддон хоть и позволяет делать экспорт в формат ЧН/ЗП, но у него осталось ограничение на 2 веса, как в формате ТЧ. Аддон пока не поддерживает 3 и 4 веса на вершину. В итоге после экспорта, вершина, которая имела 4 веса, будет иметь 2.

Проверь вершины пальцев и посмотри, сколько весов они имеют. Если больше 2, то причина в этом.
![02](https://github.com/PavelBlend/blender-xray/assets/7983249/1c9e6975-2420-44fd-a94c-613d4d24bf9d)

Вот сделал тестовую модель, экспортировал, а потом опять импортировал:
![01](https://github.com/PavelBlend/blender-xray/assets/7983249/49e5a3fa-7078-4070-adc8-e872bb31ede9)

Отличия заметны только во время анимации. В позе покоя все вершины на месте. У тебя ведь так же?

Я создал issue #723 , чтобы в будущем сделать поддержку 3 и 4 весов.

**Comment by Max1m1661** [comment]
@PavelBlend всё верно.

В позе покоя никаких отличий; некоторые вершины имеют 3-4 веса.

Спасибо за ответ и объяснение, удачи в разработке.

**Comment by lappaduch** [comment]
@PavelBlend Hi, is there a plan or is it possible to import particles? I can open then particles.xr using ParticleEditor in SDK 0.7, but I would prefer doing it in blender and your tool so far is very powerful. Thanks

**Reply by PavelBlend** [reply]
@lappaduch hello. I have no plans to add support for particles.xr

**Comment by ghost** [comment]
Привет, вопрос.
В блендере при экспорте модель в omf 

Выдает лог

bpy.ops.xray_export.omf_file(filepath="C:\\Users\\User\\Desktop\\stalker_neytral_balon_8.omf", export_mode='OVERWRITE', export_motions=True, export_bone_parts=False, high_quality=False)

**Comment by PavelBlend** [comment]
@Stalker-Vernite привет. Это в окне Info?:
![01](https://github.com/PavelBlend/blender-xray/assets/7983249/021826dd-7fa7-4eba-8416-4917319fccc5)

Если да, то это не ошибка. В это окно блендер печатает все запущенные операторы и изменённые параметры. Главное, чтобы примерно таких окон со стеком вызовов питона не появлялось:
![1](https://private-user-images.githubusercontent.com/75882664/295645460-4e1b4a66-6605-41ab-8920-7ed27f09e940.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMzg4OTcsIm5iZiI6MTcxMTEzODU5NywicGF0aCI6Ii83NTg4MjY2NC8yOTU2NDU0NjAtNGUxYjRhNjYtNjYwNS00MWFiLTg5MjAtN2VkMjdmMDllOTQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIyVDIwMTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxNjAzNGZjMmI2ZGViODI0YTZlNTE0MWY0OWYyYTdiZTE1NGVmYThkOWQ1ZTMyNTEyMjdiMDJjZTY4Y2IyYTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-xzjY_ZRrTbrTv9nIXBURbViQDvJ7sNdrs2CtOp2_Ec)

И чтобы не появлялись такого плана известные проблемы и предупреждения:
![2](https://private-user-images.githubusercontent.com/7983249/296232711-d5fa08e4-cc68-4024-9d14-336c14457e72.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMzg5ODgsIm5iZiI6MTcxMTEzODY4OCwicGF0aCI6Ii83OTgzMjQ5LzI5NjIzMjcxMS1kNWZhMDhlNC1jYzY4LTQwMjQtOWQxNC0zMzZjMTQ0NTdlNzIuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjJUMjAxODA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWE5N2NjZjdmMzdiYTQ3M2YyMzNkNDZiOWFlZjhkNjJhYjZiNjhkMTNiNjk0NGRlN2E5YTBjMzczNmU3MWFkNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.xBHmJuhB-3LjfozswudObsbi41Mmd3GtgDVZlOhPv-s)

**Comment by ghost** [comment]
Кхм... Нет
 Вот окно
![2024-03-23_13-38-34](https://github.com/PavelBlend/blender-xray/assets/164559960/0140fec6-a0ff-4d1e-baf6-8908326ba96a)

Нажимаю export/omf И лог

bpy.ops.xray_export.omf_file(filepath="D:\\X-Ray SDK\\level_editor\\gamedata\\meshes\\stalker_neytral_balon_8.omf", export_mode='OVERWRITE', export_motions=True, export_bone_parts=False, high_quality=False)

**Comment by PavelBlend** [comment]
@Stalker-Vernite а в каком окне лог? Покажи скриншот лога.

**Comment by ghost** [comment]
![2024-03-23_14-13-16](https://github.com/PavelBlend/blender-xray/assets/164559960/c05c529e-0f5f-4ab2-9604-37d55a29fc36)

**Reply by PavelBlend** [reply]
@Stalker-Vernite это окно `Info`, как у меня на первом скриншоте. Просто у тебя оно открыто в отдельном окне. В этом журнале оповещения действий, которые совершает блендер. Так же могут отображаться ошибки с красной иконкой с крестиком. У тебя не созданы группы костей, судя по ошибке `Armature Object nas no bone groups`. Их нужно создать либо импортировать. Вот я объяснял, как их импортировать:
https://github.com/PavelBlend/blender-xray/issues/344#issuecomment-917685466

**Reply by ghost** [reply]
Спасибо большое.

Еще один вопрос.

Я вот удалил все анимации из окна вот этого
![2024-03-23_17-08-08](https://github.com/PavelBlend/blender-xray/assets/164559960/cea26335-9209-45e4-8185-f7453b2bbda0)

Потом импортирую анимации сталкеров в формате omf

При экспорте пишет в формат omf

Не все кости арматуры имеют группу костей (bone part): "stalker_neytral_balon_8.object"

Какие кости имеют ввиду?

Которые  дочерные объекты в самой моделе.

Я не правильно объясняю, вот это окно они имеют ввиду где кости?

![2024-03-23_17-10-48](https://github.com/PavelBlend/blender-xray/assets/164559960/3250ff39-9587-4064-b5a6-44c3edf317f5)

**Comment by PavelBlend** [comment]
@Stalker-Vernite для более подробной информации, нужно открыть окно `Text Editor` и в списке выбрать лог с именем `xray_log`. В скобках будут указаны кости, которые не имеют группу костей.
![01](https://github.com/PavelBlend/blender-xray/assets/7983249/06bed673-5286-4fb2-b334-8a280c107d8d)

Группы костей находятся в окне Properties (панель Bone Groups). Нужно включить режим Pose Mode у арматуры, выделить кости, выбрать группу костей (у меня на скриншоте это группа с именем default) и нажать кнопку assign, чтобы назначить выбранные кости к выбранной группе кости.
![02](https://github.com/PavelBlend/blender-xray/assets/7983249/e90a82f3-9fce-42bc-ae9a-208668166bb2)

**Reply by ghost** [reply]
Спасибо тебе огромное! Все подробно объяснил)

А что делать если у тебя 3 группы костей? То к какой именно назначать?

Каждую выделять и назначивать?

А так спасибо тебе огромное!

Извиняюсь что постоянно задаю вопросы глупые.

Но возможен ли такой вариант что я выделил все кости  в режиме позы, и назначил в группе костей все.

И в любом случае пишет в логе что все кости в логе все равно не назначены группы костей.

Я все делал как было показано на твоем скриншоте выше.

**Reply by PavelBlend** [reply]
Смотря какая модель. Если модель сталкера, то нужно настроить группы так же, как и в оригинальных моделях. Кости рук и туловища к одной группе, кости головы к другой, кости ног - к третьей. Импортирую модель и игры и потом импортируй boneparts из omf и посмотри как настроено.

**Reply by ghost** [reply]
Спасибо тебе большое! Все работает! Сильно мне помог! Правда были небольшие траблы.

1. В блендере где была временная шкала было выставлено значение 70, при загрузке анимации в AE Значение это каким то образом сбивалось, и место 70 секунд отыгрывалось только 42.
2. И еще один вопрос, а в самом плагине можно ли переименовать кости? Вернее я знаю прекрасно что их можно, просто забыл где именно. Напомни пожалуйста.
3. И еще один вопрос, можно ли каким то образом привязать камеру допустим к игроку в блендере, и что бы создавалось ощущение что камера идет с игроком. Возможно ли такое?? Или это уже по части движка.

**Comment by ghost** [comment]
По поводу первого вопроса, вот скриншот, на нем видно что всего 62 секунды анимации. А в блендере на таймплане 70 секунд.[Снимок](https://github.com/PavelBlend/blender-xray/assets/164559960/78c0f2a7-2f66-42dc-9c37-039a2e394b4d)

**Comment by PavelBlend** [comment]
@Stalker-Vernite 
1 - у тебя анимация в сдк начинается с 1 кадра. В блендере тоже? Конечный кадр может изменится, если начало смещено. При экспорте аддон начальный кадр делает нулевым. Главное, что длина анимации сохранилась. А то, что ключи смещены, ничего страшного. Вот ключи до экспорта (начало - 20, конец - 60, длина - 41):
![01](https://github.com/PavelBlend/blender-xray/assets/7983249/cc3e6a34-293b-4dd8-87b0-1b3f575aae3b)

А вот после экспорта (начало - 0, конец - 40, длина 41):
![02](https://github.com/PavelBlend/blender-xray/assets/7983249/4855fd1e-b012-4189-bfb7-d154aca7fe07)

2 - в плагине кости переименовывать нельзя. Эта возможность является стандартной возможностью блендера. Имя меняется здесь:
![03](https://github.com/PavelBlend/blender-xray/assets/7983249/97fa74c7-dece-4bea-afea-ad4b8a130f77)

3 - в блендере можно, но это поведение камеры будет только внутри блендера и экспортировать его в игру нельзя. Чтобы в игре камера была такой, можно включить режим от третьего лица. Нужно в игре открыть консоль клавишей '~' (`ё`) и в консоли вписать команды:

`bind cam_1 kF1`
`bind cam_2 kF2`
`bind cam_3 kF3`
`bind cam_4 kF4`

Потом можно переключать режимы камеры с помощью клавиш `F1-4`. Если нужна другая камера, то нужно править движок.

**Comment by ghost** [comment]
Привет. А ты не знаешь как привязать к примеру болт к рукам? Просто сейчас делаю  анимацию для гайки, вот и не получается привязать гайку к модели. Пробовал через ctrl+p но толку 0

**Comment by PavelBlend** [comment]
@Stalker-Vernite Привет. В одном сообщении сложно объяснить. Поищи уроки на ютубе по констрейнту Child of для костей. Нужно болт привязать этим констрейнтом к руке. Хотя можно и не привязывать, если скелет изменить. Всё зависит от скелета, какая иерархия костей у него. И зависит от версии игры: ТЧ или ЧН/ЗП.

**Comment by ghost** [comment]
@PavelBlend  Спасибо огромное) Разобрался. А вот теперь еще один вопрос. Как достать анимации из motion reference

К примеру я хочу их достать, и как либо править, но я вот не могу
 
![2024-04-30_15-20-22](https://github.com/PavelBlend/blender-xray/assets/164559960/493ed269-2d65-4afb-9506-86121e204a57)

**Reply by PavelBlend** [reply]
@Stalker-Vernite нужно импортировать либо omf, либо skls файл из gamedata\meshes\weapons\bolt\wpn_bolt_hud_animation. Отредактировать анимации и обратно сделать экспорт в skls/omf.

skls файл можно получить из omf с помощью конвертера бардака (https://github.com/abramcumner/xray_re-tools)

**Comment by Sardinky** [comment]
Здравствуйте. Пробую экспортировать .omf, но выдается ошибка "Armature object has no bone groups". Делаю так, загружаю ogf, выделяю скелет, импортирую omf (все анимации соответсвуют костям и работают в игре), и тут же экспортирую omf, даже без изменения самой анимации. И даже так выдется ошибка. Как экспортировать omf?

**Reply by PavelBlend** [reply]
@Sardinky Привет.

Нужно, чтобы у арматуры были группы костей.

В СДК это называется bonepart. У оружия/мутантов должен быть один bonepart default. У сталкеров должно быть три bonepart: legs, torso, head.

В блендере 3.6 и более старых boneparts - это Bone Groups.
![01](https://github.com/user-attachments/assets/c43ae40f-75b3-4fbd-9c01-a8473ce4fc39)

В блендере 4.0 и выше, boneparts - это коллекции костей.
![04](https://github.com/user-attachments/assets/97a8b878-e2e7-42c6-9348-5e13d57464e5)

По идее их не нужно создавать вручную, если ты импортировал omf. Но почему-то они не создались во время импорта omf. Какие параметры omf импорта у тебя были установлены? Нужно, чтобы был включен параметр Import Boneparts:
![02](https://github.com/user-attachments/assets/e1049311-8c79-4383-b6e7-98f3b9f89fd9)

Либо можно в настройках аддона указать meshes folder, тогда boneparts аддон импортирует во время импорта ogf (он найдёт нужный omf файлы и, во время импорта ogf, прочитает bonparts из omf):
![03](https://github.com/user-attachments/assets/1afd07a3-743a-4074-89b3-565c9d979f2a)

**Reply by Sardinky** [reply]
Огромное спасибо! Сработало

**Comment by AkciizzZ** [comment]
Здравствуйте, Павел!

Столкнулся с проблемой по экспорту анимаций *.skl для оружия.

Мой порядок действий:
1. Подгрузил модели рук и автомата;
2. Создал риг по вашим урокам с ютуба;
3. Для теста написал простейшую анимацию выстрела.

Теперь не могу экспортировать анимацию для рук и для оружия, так как блендер пишет ошибку "**Selected objects have no actions**".

В View port включен режим Pose, а во втором окне Action Editor. Пробовал включать просто Dope Sheet, но разницы нет никакой. 
Выделял и анимации, и кости, и объекты в правом верхнем браузере и все равно выходит эта ошибка.

Помогите, пожалуйста, разобраться, где я накосячил.

![image](https://github.com/user-attachments/assets/23eb51a3-4f54-49a1-af1f-9b997405e25c)

Спасибо!

**Reply by PavelBlend** [reply]
@AkciizzZ Привет.

У объекта заполнен этот список?
![01](https://github.com/user-attachments/assets/5d42ece7-4fb7-4c18-9210-7b5230bc85a6)

Он не должен быть пустым.

**Reply by AkciizzZ** [reply]
О, получилось экспортировать) Действительно, этот список оказался пустым. Добавил анимацию и после этого удалось выгрузить ее.
Спасибо!!!

Однако возникла новая проблемка. Для теста загрузил модель рук и оружия в 3ds Max и их новые анимации. В позу встают правильно, но почему-то у кости **lead gun** нет ниодной анимации - все ключи пустые, хотя в Blender ключи есть. 
По итогу, двигаются руки, затвор и спусковой крючок, а сам автомат стоит на месте. Если  сконнектить **wpn_body** к **lead gun**, то все равно результат нулевой - автомат стоит. Кроме того у **wpn_body** по какой-то причине есть два ключа, которых нет в Blender.

У **lead gun** три ключа, не считая начало и конец:
![image](https://github.com/user-attachments/assets/2aaceabb-0527-4903-a91f-2b396d026e1c)

**wpn_body** - неимеет никаких ключей:
![image](https://github.com/user-attachments/assets/66195fbc-d27e-4309-b58b-74fda72e5992)

![image](https://github.com/user-attachments/assets/f5be3975-2aa7-4df9-9a4b-25e88fb63dd2)

![image](https://github.com/user-attachments/assets/155f5b6a-8f55-4ba5-bffc-c247f601c1f3)

Опять-таки, даже если сконнектить **wpn_body** к **lead gun**, то эффекта ноль, так как **lead gun** без ключей(.

**Reply by PavelBlend** [reply]
@AkciizzZ можешь скинуть мне blend файл, чтобы я у себя смог посмотреть и проверить?

**Reply by AkciizzZ** [reply]
[Для теста.zip](https://github.com/user-attachments/files/20723450/default.zip)

Да, конечно, вот архив с проектом и моделями.

**Reply by PavelBlend** [reply]
@AkciizzZ проверил. У кости `lead_gun` выключен параметр `Exportable` (экспортируемая кость). Нужно включить.
![01](https://github.com/user-attachments/assets/7059d73d-38b0-4fb0-b7a7-5ccba02c4038)

У кости `wpn_body` появляются ключи, потому-что кость движется с помощью констрейнта `Copy Transforms`. Экспортёру не важно, сколько ключей у кости, а важно, движется кость или нет. Правда у меня не два ключа, а больше:
![02](https://github.com/user-attachments/assets/63e2742a-f72d-4602-b8da-a9454d11d3dd)

**Reply by AkciizzZ** [reply]
Ура! Заработало!!! Спасибо большое)))
Теперь экспортировалось как надо. Сконнектил в 3ds max `wpn_body` к `lead gun` и получилось все так, как надо. 
Еще раз спасибо за помощь!)))

**Comment by AkciizzZ** [comment]
Павел, добрый день! 
Вопросик есть, помогите пожалуйста.

Подгрузил в блендер руки из ЗП и анимации для АС Вала. Цель была такова, чтобы из анимации `idle` сделать новую `idle_aim` без смещения в бок. 
Взял `idle` скопировал в новую анимацию и для всех костей стер ключи, кроме начала и конца. 
Произвел экспорт в `val_idle_aim2_001.skl`.
Загрузил в СДК 0.7 модель рук, в разделе Motoins стёр все ссылки на пакеты с анимациями для других оружий и детекторов.
Далее загрузил через `Append` оригинальный skls (конвертированный OMF), а затем добавил новенький `val_idle_aim2_001.skl`. 
В СДК все показывается хорошо, однако, стоит мне сделать экспорт в OMF, то игра выдает вылет, причем, по памяти, что странно.
Брал и другие СДК для экспорта OMF нового пакета анимаций рук для Вала и все равно игра уходит в ошибку.
Не подскажите, что я делаю не так?

В СДК все хорошо:
<img width="1090" height="1017" alt="image" src="https://github.com/user-attachments/assets/aa141a7f-e490-4fc7-8dae-bf26f42733f0" />

**Reply by AkciizzZ** [reply]
Если поподробней.
`wpn_hand_val_hud_animation.omf` перевожу в` wpn_hand_val_hud_animation.skls`, который в себе содержит эти анимации:
<img width="369" height="152" alt="image" src="https://github.com/user-attachments/assets/5ad52992-199d-443f-a78d-b3c85f7bccce" />
Я хочу добавить новую из брендера `val_idle_aim2_001.skl`, а затем экспортировать в OMF новый `wpn_hand_val_hud_animation.omf`.

**Reply by AkciizzZ** [reply]
Но игра с новым `wpn_hand_val_hud_animation.omf` вылетает. Почему-то при экспорте из СДК все ломается.

**Reply by AkciizzZ** [reply]
Ура! Отбой вопроса.
Загрузил в Блендер модель рук и `wpn_hand_val_hud_animation.omf`, сделал новую `val_idle_aim2_001.skl` и экспортировал как OMF с объединением с иходным `wpn_hand_val_hud_animation.omf`. В конфиге поменял название анимации и все поехало)
Результат на видео: https://www.youtube.com/watch?v=JnhTRCWz8i0

**Comment by Dmitriy-Iassenev** [comment]
Возникла проблема: в интернете отсутствуют исходники релизных уровней для «Тени Чернобыля». Конечно, существуют исходники, но они датированы ноябрем 2006 года. Поэтому я решил прогнать обычный уровень «Кордон» через Blender, чтобы исключить возможные проблемы, связанные с декомпиляцией.

Однако у меня возникла проблема: я не знаю, как прогнать локацию через Blender и преобразовать её в формат SDK 0.4.
